From 0ae4bc2052218523423bae070dda3f5a31d50001 Mon Sep 17 00:00:00 2001
From: Carlos Antonio da Silva <carlosantoniodasilva@gmail.com>
Date: Sun, 1 Aug 2010 15:52:10 -0300
Subject: [PATCH] Fix label form helper to use I18n and html options, without the need of 'nil' text param:

  Before: f.label :title, nil, :class => 'title'
  After : f.label :title, :class => 'title'
---
 actionpack/lib/action_view/helpers/form_helper.rb |    7 ++++---
 actionpack/test/template/form_helper_test.rb      |   11 +++++++++--
 2 files changed, 13 insertions(+), 5 deletions(-)

diff --git a/actionpack/lib/action_view/helpers/form_helper.rb b/actionpack/lib/action_view/helpers/form_helper.rb
index b0af836..54bb72b 100644
--- a/actionpack/lib/action_view/helpers/form_helper.rb
+++ b/actionpack/lib/action_view/helpers/form_helper.rb
@@ -215,7 +215,7 @@ module ActionView
       #    ...
       #   <% end %>
       #
-      # If your resource has associations defined, for example, you want to add comments 
+      # If your resource has associations defined, for example, you want to add comments
       # to the post given that the routes are set correctly:
       #
       #   <%= form_for([@document, @comment]) do |f| %>
@@ -583,8 +583,9 @@ module ActionView
       #     'Accept <a href="/terms">Terms</a>.'
       #   end
       def label(object_name, method, content_or_options = nil, options = nil, &block)
-        if block_given?
-          options = content_or_options if content_or_options.is_a?(Hash)
+        content_is_options = content_or_options.is_a?(Hash)
+        if content_is_options || block_given?
+          options = content_or_options if content_is_options
           text = nil
         else
           text = content_or_options
diff --git a/actionpack/test/template/form_helper_test.rb b/actionpack/test/template/form_helper_test.rb
index 9ddfa13..749abb5 100644
--- a/actionpack/test/template/form_helper_test.rb
+++ b/actionpack/test/template/form_helper_test.rb
@@ -110,6 +110,13 @@ class FormHelperTest < ActionView::TestCase
     I18n.locale = old_locale
   end
 
+  def test_label_with_locales_and_options
+    old_locale, I18n.locale = I18n.locale, :label
+    assert_dom_equal('<label for="post_body" class="post_body">Write entire text here</label>', label(:post, :body, :class => 'post_body'))
+  ensure
+    I18n.locale = old_locale
+  end
+
   def test_label_with_for_attribute_as_symbol
     assert_dom_equal('<label for="my_for">Title</label>', label(:post, :title, nil, :for => "my_for"))
   end
@@ -603,7 +610,7 @@ class FormHelperTest < ActionView::TestCase
 
   def test_form_for_with_symbol_object_name
     form_for(@post, :as => "other_name", :html => { :id => 'create-post' }) do |f|
-      concat f.label(:title)
+      concat f.label(:title, :class => 'post_title')
       concat f.text_field(:title)
       concat f.text_area(:body)
       concat f.check_box(:secret)
@@ -611,7 +618,7 @@ class FormHelperTest < ActionView::TestCase
     end
 
     expected =  whole_form("/posts/123", "create-post", "other_name_edit", :method => "put") do
-      "<label for='other_name_title'>Title</label>" +
+      "<label for='other_name_title' class='post_title'>Title</label>" +
       "<input name='other_name[title]' size='30' id='other_name_title' value='Hello World' type='text' />" +
       "<textarea name='other_name[body]' id='other_name_body' rows='20' cols='40'>Back to the hill and over it again!</textarea>" +
       "<input name='other_name[secret]' value='0' type='hidden' />" +
-- 
1.7.2

